c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -3166,6 +3235,315 @@ namespace TransformStream {
+    return true;
+  }
+
+  // Steps 1-5 of the transform algorithm, and 1-4 of the flush algorithm.
+  bool deflate_chunk(JSContext* cx, HandleObject self, HandleValue chunk, bool finished) {
+    z_stream* zstream = state(self);
+
+    if (!finished) {
+      // 1.  If _chunk_ is not a `BufferSource` type, then throw a `TypeError`.
+      // Step 2 of transform:
+      size_t length;
+      uint8_t* data = value_to_buffer(cx, chunk, "CompressionStream transform: chunks", &length);
+      if (!data) {
+        return false;
+      }
+
+      if (length == 0) {
+        return true;
+      }
+
+      // 2.  Let _buffer_ be the result of compressing _chunk_ with _cs_'s format and context.
+      // This just sets up step 2. The actual compression happen in the `do` loop below.
+      zstream->avail_in = length;
+      zstream->next_in = data;
+    } else {
+      // Step 1 of flush:
+      // 1.  Let _buffer_ be the result of compressing an empty input with _cs_'s format and
+      //     context, with the finish flag.
+      // This just sets up step 2. The actual compression happen in the `do` loop below.
+      zstream->avail_in = 0;
+      zstream->next_in = nullptr;
+    }
+
+    RootedObject controller(cx, TransformStream::controller(transform(self)));
+
+    // Steps 3-5 of transform are identical to steps 2-4 of flush, so numbers below refer to
+    // the former for those.
+    // Also, the compression happens in potentially smaller chunks in the `do` loop below, so
+    // the three steps are reordered and somewhat intertwined with each other.
+
+    uint8_t* buffer = output_buffer(self);
+
+    do {
+      // 4.  Split _buffer_ into one or more non-empty pieces and convert them into `Uint8Array`s.
+      // 5.  For each `Uint8Array` _array_, enqueue _array_ in _cs_'s transform.
+      // This loop does the actual compression, one output-buffer sized chunk at a time, and then
+      // creates and enqueues the Uint8Arrays immediately.
+      zstream->avail_out = BUFFER_SIZE;
+      zstream->next_out = buffer;
+      int err = deflate(zstream, finished ? Z_FINISH : Z_NO_FLUSH);
+      if (!((finished && err == Z_STREAM_END) || err == Z_OK)) {
+        JS_ReportErrorASCII(cx, "CompressionStream transform: error compressing chunk");
+        return false;
+      }
+
+      size_t bytes = BUFFER_SIZE - zstream->avail_out;
+      if (bytes) {
+        RootedObject out_obj(cx, JS_NewUint8Array(cx, bytes));
+        if (!out_obj) {
+          return false;
+        }
+
+        bool is_shared;
+        JS::AutoCheckCannotGC nogc;
+        uint8_t* out_buffer = JS_GetUint8ArrayData(out_obj, &is_shared, nogc);
+        memcpy(out_buffer, buffer, bytes);
+        RootedValue out_chunk(cx, ObjectValue(*out_obj));
+        if (!TransformStreamDefaultController::Enqueue(cx, controller, out_chunk)) {
+          return false;
+        }
+      }
+
+      // 3.  If _buffer_ is empty, return.
+    } while (zstream->avail_out == 0);
```

Comment:
I'm not sure I understand the loop condition `zstream->avail_out == 0` here: does this mean that we keep invoking `deflate` as long as it *fully* fills the output chunk? Is it possible for `deflate` to fill, say, all but a few bytes of the output (maybe there isn't enough space to fully encode a symbol) and then return while there is still more available input?

Said another way, what prevents the loop condition from being `while (zstream->avail_in > 0)` ? Or `while ((bytes > 0) || (zstream->avail_in > 0))` (i.e., as long as either there is more input to process, or this invocation produced some output)?

Reply:
Very good question! The trivial answer to "why did I do it this way" is that that's how Gecko and [Chromium](https://chromium.googlesource.com/chromium/src/+/e8564ac44910253b6daada89fd67405c3a886f57/third_party/blink/renderer/modules/compression/deflate_transformer.cc#99) (in this case for exactly the same purpose) use zlib.

I added a comment explaining this and referring to the zlib docs which [explicitly state](https://searchfox.org/mozilla-central/rev/87ecd21d3ca517f8d90e49b32bf042a754ed8f18/modules/zlib/src/zlib.h#319-324) that this is what one should do.